File: 백준/Silver/10971. 외판원 순회 2/외판원 순회 2.cc
#include<iostream>

using namespace std;

int N;
int arr[12][12] = { 0, };
int visit[12] = { 0, };
int order[12] = { 0, };
int sol = 100000000;

void dfs(int now_idx, int count, int total) {
	if (count == N) {
		if (arr[now_idx][order[1]] != 0) {
			total += arr[now_idx][order[1]];
			if (total < sol) {
				sol = total;
			}
		}
		return;
	}

	for (int i = 1; i <= N; i++) {
		if (count == 0) {
			visit[i] = 1;
			order[count + 1] = i;
			dfs(i, count + 1, total);
			visit[i] = 0;
		}
		else if (arr[now_idx][i] != 0) {
			if (visit[i] == 0) {
				visit[i] = 1;
				order[count + 1] = i;
				dfs(i, count + 1, total + arr[now_idx][i]);
				visit[i] = 0;
			}
		}
	}
}

int main() {


	cin >> N;

	for (int i = 1; i <= N; i++) {
		for (int j = 1; j <= N; j++) {
			cin >> arr[i][j];
		}
	}

	dfs(0, 0, 0);

	cout << sol;


	return 0;
}

File: 백준/Silver/10971. 외판원 순회 2/외판원 순회 2.java
import java.io.*;
import java.util.Arrays;

public class Main {
    public static int[] check;  // 지나간 도시 체크
    public static int sol = 100000000;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int N = Integer.parseInt(br.readLine());
        int[][] arr = new int[N][N];
        check = new int[N];

        // 도시 사이의 이동비용 배열 세팅
        for (int i = 0; i < N; i++) {
            String[] str = br.readLine().split(" ");
            for (int j = 0; j < N; j++) {
                arr[i][j] = Integer.parseInt(str[j]);
            }
        }

        check[0] = 1;
        dfs(arr, 0, 1, 0);

        bw.write(sol+"");
        bw.flush();
        bw.close();

    }

    public static void dfs(int[][] arr, int spot, int count, int total) {
        // 마지막 도시에 도착했을때가 탈출조건
        if (count == arr.length && arr[spot][0] != 0) {
            // 마지막 도시에서 첫번째 도시로 이동할때 비용 추가
            total += arr[spot][0];
            // 최소 비용 체크
            if (sol > total) {
                sol = total;
            }
            return;
        }

        for (int i = 0; i < arr.length; i++) {
            if (check[i] == 0 && arr[spot][i] != 0) {
                check[i] = 1;
//                System.out.println("node= " + i + "  count= " + (count+1) + " total= " + (total+arr[spot][i]));
                dfs(arr, i, count + 1, total+arr[spot][i]);
//                System.out.println("                   이전으로  count= " + count);
                check[i] = 0;
            }
        }
    }
}